### PR TITLE
[Serializer] throw more specific exceptions

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20534
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by fabpot** [comment]
👍

File: src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php [hunk 1]
```diff
@@ -81,12 +81,12 @@ public function supportsNormalization($data, $format = null)
...
      *
      * @throws InvalidArgumentException
-     * @throws UnexpectedValueException
+     * @throws NotNormalizableValueException
      */
     public function denormalize($data, $class, $format = null, array $context = array())
     {
         if (!preg_match('/^data:([a-z0-9][a-z0-9\!\#\$\&\-\^\_\+\.]{0,126}\/[a-z0-9][a-z0-9\!\#\$\&\-\^\_\+\.]{0,126}(;[a-z0-9\-]+\=[a-z0-9\-]+)?)?(;base64)?,[a-z0-9\!\$\&\\\'\,\(\)\*\+\,\;\=\-\.\_\~\:\@\/\?\%\s]*\s*$/i', $data)) {
-            throw new UnexpectedValueException('The provided "data:" URI is not valid.');
+            throw new NotNormalizableValueException('The provided "data:" URI is not valid.');
```
Comment (dawehner) [inline src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php]: IMHO this example here should actually continue to throw the old exception. This is decodeable, but just not valid in the domain. I guess this example would be a 422 instead?

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php [hunk 2]
```diff
@@ -79,7 +79,7 @@ public function denormalize($data, $class, $format = null, array $context = arra
 
             $dateTimeErrors = \DateTime::class === $class ? \DateTime::getLastErrors() : \DateTimeImmutable::getLastErrors();
 
-            throw new UnexpectedValueException(sprintf(
+            throw new NotNormalizableValueException(sprintf(
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: Maybe can we set a specific exception code for the different class of exceptions to be able to easily detect if it's an invalid data: URI, or an invalid date and so on?

**Comment by fabpot** [comment]
@xabbuh Any feedback on the comments?

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 2 above)
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: I have thought about that too. The reason I did not add it is consistency. We would only be able to add codes for Symfony's own normalizer. However, it's quite common to add your own normalizers in libraries/projects. As a consequence exception codes would only be useful when they are issued by built-in normalizers. I don't really like that. What do you think?

File: src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php (hunk 1 above)
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php]: But the same is true for the other places where we throw a `NotNormalizableValueException` in this PR, isn't it?

**Comment by fabpot** [comment]
@xabbuh What about this one?

**Comment by xabbuh** [comment]
I think I need @dunglas to answer my comments. :)

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 2 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: We can maybe reserve a range for app codes?

**Review by dunglas — APPROVED** [review]

**Comment by fabpot** [comment]
So, now that @dunglas approved this PR, @xabbuh can you rebase it so that we can merge it? Thanks.

**Comment by xabbuh** [comment]
PR rebased

**Review by ogizanagi — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by dawehner** [comment]
Thank you for solving this problem!

**Pull request merged**